sulong/projects/com.oracle.truffle.llvm.nodes/src/com/oracle/truffle/llvm/nodes/intrinsics/multithreading/LLVMPThreadIntrinsics.java

```diff
@@ -0,0 +1,847 @@
+                CompilerDirectives.transferToInterpreterAndInvalidate();
+                store = LLVMLanguage.getLLVMContextReference().get().getNodeFactory().createStoreNode(LLVMInteropType.ValueKind.I32);
+            }
+            // store type in attr var
+            if (type == null) {
+                type = new Integer(0);
+            }
+            store.executeWithTarget(attr, type);
+            // TODO: return with error if type not in {0, 1, 2}
+            // look up fitting error code
+            return 0;
+        }
+    }
+
+    @NodeChild(type = LLVMExpressionNode.class)
+    public abstract static class LLVMPThreadMutexDestroy extends LLVMBuiltin {
+        @Specialization
+        protected int doIntrinsic(VirtualFrame frame, Object mutex) {
+            long mutexAddress = ((LLVMNativePointer) mutex).asNative();
+            LLVMLanguage.getLLVMContextReference().get().mutexStorage.remove(mutexAddress);
+            return 0;
+        }
+    }
+
+    @NodeChild(type = LLVMExpressionNode.class)
+    @NodeChild(type = LLVMExpressionNode.class)
+    public abstract static class LLVMPThreadMutexInit extends LLVMBuiltin {
+        @Child
+        LLVMLoadNode read = null;
+
+        @Specialization
+        protected int doIntrinsic(VirtualFrame frame, Object mutex, Object attr) {
+            if (read == null) {
+                CompilerDirectives.transferToInterpreterAndInvalidate();
+                read = LLVMLanguage.getLLVMContextReference().get().getNodeFactory().createLoadNode(LLVMInteropType.ValueKind.I32);
+            }
+            // we can use the address of the native pointer here, bc a mutex
+            // must only work when using the original variable, not a copy
+            // so the address may never change
+            long mutexAddress = ((LLVMNativePointer) mutex).asNative();
+            Object mutObj = LLVMLanguage.getLLVMContextReference().get().mutexStorage.get(mutexAddress);
+            int attrValue = 0;
+
+            if (!((LLVMPointer) attr).isNull()) {
+                attrValue = (int) read.executeWithTarget(attr);
+            }
+            Mutex.Type mutexType = Mutex.Type.DEFAULT_NORMAL;
+            if (attrValue == 1) {
+                mutexType = Mutex.Type.RECURSIVE;
+            } else if (attrValue == 2) {
+                mutexType = Mutex.Type.ERRORCHECK;
+            }
+            if (mutObj == null) {
+                LLVMLanguage.getLLVMContextReference().get().mutexStorage.put(mutexAddress, new Mutex(mutexType));
+            }
+            return 0;
+        }
+    }
+
+    @NodeChild(type = LLVMExpressionNode.class)
+    public abstract static class LLVMPThreadMutexLock extends LLVMBuiltin {
+        @Specialization
+        protected int doIntrinsic(VirtualFrame frame, Object mutex) {
+            long mutexAddress = ((LLVMNativePointer) mutex).asNative();
+            Mutex mutexObj = (Mutex) LLVMLanguage.getLLVMContextReference().get().mutexStorage.get(mutexAddress);
```

Comment:
Is there any reason why the `mutexStorage` needs to use a `long` as a key rather than an `LLVMPointer` or `Object`. This code will likely not work in Sulong's sandboxed mode. There you'd have only `LLVMManagedPointer`. In general, the load and store nodes produced by Sulong can handle these, so wherever possible I'd like you to use `LLVMPointer` rather than `LLVMNativePointer`.